prime-router/src/main/kotlin/cli/tests/BasicTests.kt

```diff
@@ -115,126 +137,194 @@ class End2EndUniversalPipeline : CoolTest() {
+                expectedReceiver,
+                sender,
+                blobContainerMetadata
             )
+
+            // Compare the actual file content with the expected content
+            val expectedFormat = Report.Format.valueOfFromExt(expectedReceiver.translation.type)
+            passed = passed and CompareData().compare(
+                expectedByteArray.inputStream(),
+                actualByteArray.inputStream(),
+                expectedFormat,
+                null
+            ).passed
+
+            if (passed) {
+                good("The contents of $actualFilename matches the expected data")
+            } else {
+                bad("***$name test FAILED***: The contents of $actualFilename did not match the expected contents")
+            }
         }
+
         return passed
     }
+
+    /**
+     * Searches a submission history json response for the external filename that matches the expected receiver.
+     * Will fail if the expected json objects are not populated.
+     * @param response must be a response from the submission history endpoint for json lookup
+     * @param expectedReceiver receiver for which to find a matching destination service
+     */
+    private fun findReportExternalName(response: String, expectedReceiver: Receiver): String {
+        var filename = ""
+
+        try {
+            val destinations = JSONObject(response).getJSONArray("destinations")
+
+            destinations.forEach { d ->
+                val destination = JSONObject(d.toString())
+
+                // Check that the destination service matches the expected receiver. Uses the first name listed.
+                if (expectedReceiver.name == destination.getString("service").removeSurrounding("\"")) {
+                    filename = destination.getJSONArray("sentReports").getJSONObject(0).getString("externalName")
+                    good("The report was received by ${expectedReceiver.name}")
+                }
+            }
+        } catch (e: org.json.JSONException) {
+            bad("***$name test FAILED***: History response is missing an element for the sent report: $e")
+        }
+
+        return filename
+    }
+
+    /**
+     * Performs the expected transforms that would happen to a file moving through the UP. If the topic is ELIMS this
+     * process will be skipped. Otherwise, the file will be converted to FHIR if necessary. Sender enrichments will be
+     * run. Then translated based on the expected receiver format.
+     * Caveats: This does not perform all transformations from the UP. Notably, batch headers/footers are absent.
+     * @param originalFile the file to transform
+     * @param expectedReceiver determines resulting data format
+     * @param sender used for the sender enrichment schema
+     * @param blobMetadata necessary for the FhirTransformer
+     * @return the transformed ByteArray in the proper data format of HL7 or FHIR
+     */
+    private fun performFileTransforms(
```

Comment:
Is this function necessary? Could you not just create the control file manually then supply that in to compare to the output? 

If there was a bug introduced to another part of the code, like maybe in `FhirToHl7Converter()`, this function wouldn't catch that, right?  To me it looks like it would make a new control file with the code that has a bug then compare that to the file made by the pipeline that also uses the code with a bug. So the files would be the same and the test would pass.

Reply:
That's a good question. Because it's using the same classes as the actual posted report, any bugs should be mirrored. Off the top of my head I think the main issue would be if there are any changes to conversion/transformation in the pipeline, or the sender/receiver, then the expected file would have to be updated.

Reply:
Discussed this in stand up today. The general choices are to generate the expected file in the test and provide extra test coverage for translation schemas or use a saved expected which provides extra test coverage for the various steps of the UP. It was decided that saved expected files were the route to go, but that the merit of generating them could be useful in future tests so this method has been preserved. https://github.com/CDCgov/prime-reportstream/pull/13679/files?w=0#diff-4025d6168d38b38532a410c9be4bcfa26e83dfb2239af85cf8569c57a77374a8R276